File: typeahead/typeahead.py
class Typeahead:

    # @param dict: A dictionary of words dict
    def __init__(self, dict):
        # do initialize if necessary
        self.word_dict = {}
        for s in dict:
            s_len = len(s)
            for i in range(s_len):
                for j in range(i, s_len):
                    word = s[i: j + 1]
                    if word not in self.word_dict:
                        self.word_dict[word] = [s]
                    elif self.word_dict[word][-1] != s:
                        self.word_dict[word].append(s)

    # @param word: a string
    # @return a list of words
    def search(self, word):
        # write your code here
        if word in self.word_dict:
            return self.word_dict[word]
        else:
            return []

File: typeahead/typeahead.cc
class Typeahead {
public:
    // @param dict: A dictionary of words dict
    Typeahead(unordered_set<string> &dict) {
        // do initialize if necessary
        for (auto &s : dict) {
            int l = s.length();
            for (int i = 0; i < l; ++i) {
                for (int j = 1; j <= l - i; ++j) {
                    string t = s.substr(i, j);
                    if (word_dict.find(t) == word_dict.end()) {
                        word_dict[t] = vector<string>{s};
                    } else if (*(word_dict[t].end() - 1) != s) {
                        word_dict[t].push_back(s);
                    }
                }
            }
        }
    }

    // @param str: a string
    // @return a set of words
    vector<string> search(string &str) {
        // Write your code here
        if (word_dict.find(str) != word_dict.end()) {
            return word_dict[str];
        } else {
            return vector<string>();
        }
    }

private:
    map<string, vector<string>> word_dict;
};

